nexus/src/app/sagas/instance_update/mod.rs

```diff
@@ -0,0 +1,2449 @@
+                    deprovision = true;
+                    // Similarly, if the active VMM was destroyed and the
+                    // instance has not migrated out of it, we must delete the
+                    // instance's network configuration. Again, if there was a
+                    // migration out, the subsequent migration-handling code
+                    // will change this to a network config update if the
+                    // instance is now living somewhere else.
+                    network_config = Some(NetworkConfigUpdate::Delete);
+                    Some(id)
+                } else {
+                    None
+                }
+            });
+
+        let destroy_target_vmm =
+            snapshot.target_vmm.as_ref().and_then(|target_vmm| {
+                if target_vmm.runtime.state == VmmState::Destroyed {
+                    // Unlink the target VMM ID.
+                    new_runtime.dst_propolis_id = None;
+                    update_required = true;
+                    Some(PropolisUuid::from_untyped_uuid(target_vmm.id))
+                } else {
+                    None
+                }
+            });
+
+        // If there's an active migration, determine how to update the instance
+        // record to reflect the current migration state.
+        if let Some(ref migration) = snapshot.migration {
+            if migration.either_side_failed() {
+                // If the migration has failed, clear the instance record's
+                // migration IDs so that a new migration can begin.
+                info!(
+                    log,
+                    "instance update (migration failed): clearing migration IDs";
+                    "instance_id" => %instance_id,
+                    "migration_id" => %migration.id,
+                    "src_propolis_id" => %migration.source_propolis_id,
+                    "target_propolis_id" => %migration.target_propolis_id,
+                );
+                new_runtime.migration_id = None;
+                new_runtime.dst_propolis_id = None;
+                update_required = true;
+                // If the active VMM was destroyed, the network config must be
+                // deleted (which was determined above). Otherwise, if the
+                // migration failed but the active VMM was still there, we must
+                // still ensure the correct networking configuration
+                // exists for its current home.
+                //
+                // TODO(#3107) This is necessary even if the instance didn't move,
+                // because registering a migration target on a sled creates OPTE ports
+                // for its VNICs, and that creates new V2P mappings on that sled that
+                // place the relevant virtual IPs on the local sled. Once OPTE stops
+                // creating these mappings, this path only needs to be taken if an
+                // instance has changed sleds.
```

Comment:
does anyone know whether this comment, which i blindly copied from its previous home in `ensure_instance_network_config`, still holds true? cc @internet-diglett, it looks like you closed the referenced issue #3107 (https://github.com/oxidecomputer/omicron/issues/3107#issuecomment-2083662330)

Reply:
Hmm, looks like the comment is still there, at least: https://github.com/oxidecomputer/omicron/blob/4cb3a232a2ee0fb222d65af41168ec72d04ffaf4/nexus/src/app/instance_network.rs#L351-L360

Reply:
@hawkw as long as the relevant records are present in the db, the rpw will ensure that the v2p mappings are applied to the correct sleds.

Reply:
The commented function eagerly triggers another run of the RPW, and also ensures that the correct NAT entries are present and the relevant RPW is triggered for that as well.

Reply:
> Once OPTE stops creating these mappings...

I'm not entirely sure what this part means, tbh.

Reply:
I'm not sure either, this comment was copied from existing code.

Reply:
Actually, reviewing it again jarred some things loose. So, with the way the Nexus RPW is currently written, Nexus will always win regardless of what OPTE / XDE does under the hood, as Nexus will remove any conflicting mappings and apply what is in the DB. So while opte#368 still needs to be address, Nexus does indeed enforce the correct mappings across all sleds.

Reply:
I think the comment is still half-good, in the sense that we still want to trigger networking RPWs on a failed migration out, even though the instance's active VMM and sled didn't change. This will be true at least until opte#368 is fixed, but even then it might still be simplest just to trigger those RPWs every time a migration is retired.

For background: if you have sleds A, B, and C, and you start running an instance on sled A, all three sleds need to know that traffic destined for the sled's virtual IPs should be directed to sled A's physical network interface. When I wrote this comment, that happened by different means for the different sleds:

* B and C get an update from Nexus that says "hey, these IPs are in an instance that's on sled A."
* A doesn't get this update. Instead, when the instance is incarnated there, sled agent creates a VNIC for the instance's network interface, and that implicitly installs (in A's OPTE instance) a mapping that says "this virtual IP is on the local host."

IIUC, the V2P RPW changes things so that the "instance is here" broadcasts go to all sleds, including sled A in this example. This is fine, but I don't think it addresses the underlying problem, which is:

* Instance starts on sled A
* A, B, and C find out (by one of the two means above) that the instance is now on sled A
* Instance starts migrating to sled B
* The VNIC setup process on sled B installs an "instance's VIPs are routed to the local host" entry in OPTE (second item above)
* The migration fails and unwinds; now someone needs to tell sled B that no, actually, the instance is back over on sled A

A periodic activation of the V2P update RPW will take care of this, but since we want to get these updated as soon as possible (having the wrong entry on sled B will prevent other instances in the same VPC that happen to be on sled B from routing traffic to the right place), we should still trigger that RPW explicitly here.

Reply:
Yeah, that's my understanding as well. I should probably adjust the comment to
just say that we're doing this to make sure this happens in a timely manner...